test/nn/test_load_state_dict.py

```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+            cls, data.size(),
+            dtype=data.dtype, layout=data.layout,
+            device=data.device, requires_grad=data.requires_grad,
+            strides=data.stride(), storage_offset=data.storage_offset())
+        return t
+
+    def __init__(self, data: torch.Tensor):
+        self._data = data
+
+    def __repr__(self):
+        return f"MyWrapperLoadTensor({self._data.__repr__()})"
+
+    # enable deepcopy
+    def new_empty(self, shape):
+        return type(self)(self._data.new_empty(shape))
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        kwargs = {} if kwargs is None else kwargs
+
+        def module_load_from(dest, src):
+            # always convert src to MyWrapperLoadTensor
+            if type(src) is torch.Tensor:
+                return MyWrapperLoadTensor(src)
+            elif type(src) is MyWrapperLoadTensor:
+                return src
+            elif type(src) is MyLoadTensor:
+                return MyWrapperLoadTensor(src.data)
+            else:
+                raise NotImplementedError(f"module_load_from(): unsupported type(src) {type(src)}")
+
+        def module_load_to(src, dest):
+            if type(dest) in {torch.nn.Parameter, torch.Tensor, MyLoadTensor, MyWrapperLoadTensor}:
+                return src
+            else:
+                raise NotImplementedError(f"module_load_to(): unsupported type(dest) {type(dest)}")
+
+        if func is torch.Tensor.module_load_from:
+            return module_load_from(*args, **kwargs)
+        elif func is torch.Tensor.module_load_to:
+            return module_load_to(*args, **kwargs)
+        else:
+            with torch._C.DisableTorchFunctionSubclass():
+                return func(*args, **kwargs)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+
+        def unwrap(t):
+            return t._data if isinstance(t, MyWrapperLoadTensor) else t
+
+        kwargs = {} if kwargs is None else kwargs
+        out = func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs))
+        if isinstance(out, torch.Tensor):
+            return MyWrapperLoadTensor(out)
+        else:
+            return out
+
+
+class TestLoadStateDictSwap(TestCase):
+    @skipIfCrossRef
+    @skipIfTorchDynamo("Can't swap with dynamo as dynamo installs weakrefs")
+    @wrapSwapTensorsTest(swap=True)
+    def test_swap_subclass(self):
+
+        def _create_model(subclass=None):
+            m = torch.nn.Linear(2, 3, bias=False)
+            m.register_buffer('buf', torch.randn(2, 3))
+            if subclass is not None:
+                m.weight = torch.nn.Parameter(subclass(m.weight))
+                m.buf = subclass(m.buf)
+            return m
+
+        def _test(module_subclass=None, sd_subclass=None):
+            m = _create_model(module_subclass)
+            sd = _create_model(sd_subclass).state_dict()
+            # deepcopy as swap_tensors might destructively modify state dict
```

Comment:
Ho that is indeed an interesting side effect...

Reply:
The current state is that 
- swap_tensors does not destructively modify state dict
-  in the case where `param` or `sd[key]` has `__torch_function__` handler for `module_load` if `id(param.module_load(sd[key])) == id(sd[key])`, `param`'s `THPVariable->cdata` will point to the same `TensorImpl` as `sd[key]`'s `THPVariable->cdata`
- I added a warning in the `nn.Module.load_state_dict` docs rearding this 

After thinking about it more, I am not sure that we want to `clone()` the `sd[key]` to preserve behavior of the non-swap path. I think it is unlikely that people use the `state_dict` after `load_state_dict` but curious for your feedback on this!